### DISCUSSION TITLE
Creating a logging module for CM3

### DISCUSSION THREAD

**Discussion by blakemcbride** [modula3/cm3#1194] [General]
Given the simplicity of building and maintaining the C back-end, I initially wondered why the other back-ends would be appealing at all. It seems there are two problems with the C back-end as follows:

1. Debugging

2. Interface with C programs

Depending on and linking with something as big, complex, and potentially volatile as GCC or LLVM seems untenable and impracticable for a small group. If this is true, I would think there would be a lot of value in solving the C back-end issues.

I propose the following:

A. Spend some time figuring out and correcting the issues revolving around linking with C programs. I would think this would be relatively straightforward.

B. Create a Modula-3 module similar to Java's Log4J. This is a logging package that would go a long way to ease development in Modula-3 without a full debugger.

None of this suggests that the other back-end be eliminated. It suggests that, for most people, a super-complex build process scares people away. Why don't we make the simple version more practical to use?

**Comment by mikanystrom** [comment]
I think I can give you a test example to work on if you want to try to get
to the bottom of the interoperability issues.

The debug issues may be more difficult but maybe can be solved by
generating auxiliary C code specifically so that you can call code that
"understands" the types from gdb.  Maybe...?  (That would be my first line
of attack if I had to try to solve the problem...)

   Mika
